matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java

```diff
@@ -0,0 +1,89 @@
+package com.matchbook.sdk.rest;
+
+import com.matchbook.sdk.core.StreamObserver;
+import com.matchbook.sdk.core.exceptions.MatchbookSDKParsingException;
+import com.matchbook.sdk.rest.configs.Parser;
+import com.matchbook.sdk.rest.configs.Serializer;
+import com.matchbook.sdk.rest.dtos.FailableRestResponse;
+import com.matchbook.sdk.rest.dtos.PartiallyFailableResponse;
+import com.matchbook.sdk.rest.readers.Reader;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+class PartiallyFailableResponseCallback<T extends FailableRestResponse, RESP extends PartiallyFailableResponse<T>>
+        extends RestResponseCallback<T, RESP> {
+
+    private static final int FAILABLE_HTTP_CODE = 400;
+
+    PartiallyFailableResponseCallback(StreamObserver<T> observer, Serializer serializer, Reader<T, RESP> reader) {
+        super(observer, serializer, reader);
+    }
+
+    @Override
+    public void onFailedResponse(InputStream responseInputStream, int responseCode) {
+        if (responseCode == FAILABLE_HTTP_CODE) {
+            onFailableResponse(responseInputStream, responseCode);
+        } else {
+            super.onFailedResponse(responseInputStream, responseCode);
+        }
+    }
+
+    private void onFailableResponse(InputStream responseInputStream, int responseCode) {
+        try (ByteArrayOutputStream responseBytes = copyBytes(responseInputStream);
+                InputStream copiedInputStream = copyInputStream(responseBytes)) {
+            try (Parser parser = serializer.newParser(copiedInputStream)) {
+                boolean isPartiallyFailedResponse = tryPartiallyFailedResponse(parser);
+                if (isPartiallyFailedResponse) {
+                    observer.onCompleted();
+                } else {
+                    secondParsing(responseBytes, responseCode);
+                }
```

Comment:
can be an error in the middle of response?

Reply:
Yes, the process will fail if the content is not the `PartiallyFailableResponse` we try to parse in the first attempt. Specifically, the first parsing attempt could either throw an exception or not be able to extract any meaningful information from the content. In both these cases, a second parse will be invoked to read a standard error content.

Honestly, I don't really the methods name I used (i.e. `tryPartiallyFailedResponse`, `secondParsing`). Any suggestion?